File: rust/tests/disallowed_change_types_test.rs
mod common;

use beachball::changefile::change_types::get_disallowed_change_types;
use beachball::types::change_info::ChangeType;
use beachball::types::package_info::{
    PackageGroupInfo, PackageGroups, PackageInfos, PackageOptions,
};
use common::{fake_root, make_package_infos_simple};

fn make_infos(name: &str) -> PackageInfos {
    make_package_infos_simple(&[name], &fake_root())
}

fn make_infos_with_disallowed(name: &str, disallowed: Vec<ChangeType>) -> PackageInfos {
    let mut infos = make_infos(name);
    infos.get_mut(name).unwrap().package_options = Some(PackageOptions {
        disallowed_change_types: Some(disallowed),
        ..Default::default()
    });
    infos
}

// Skipped TS tests (Rust uses Option<Vec> so null vs empty is handled differently):
// - "returns null if package disallowedChangeTypes is set to null"
// - "returns empty array if package disallowedChangeTypes is set to empty array"
// - "returns null if package group disallowedChangeTypes is set to null"
// - "returns empty array if package group disallowedChangeTypes is set to empty array"

// TS: "returns null for unknown package"
#[test]
fn returns_none_for_unknown_package() {
    let infos = PackageInfos::new();
    let groups = PackageGroups::new();
    let result = get_disallowed_change_types("unknown", &infos, &groups, &None);
    assert_eq!(result, None);
}

// TS: "falls back to main option for package without disallowedChangeTypes"
#[test]
fn falls_back_to_repo_option() {
    let infos = make_infos("foo");
    let groups = PackageGroups::new();
    let repo_disallowed = Some(vec![ChangeType::Major]);

    let result = get_disallowed_change_types("foo", &infos, &groups, &repo_disallowed);
    assert_eq!(result, Some(vec![ChangeType::Major]));
}

// TS: "returns disallowedChangeTypes for package"
#[test]
fn returns_package_level_disallowed() {
    let infos = make_infos_with_disallowed("foo", vec![ChangeType::Major, ChangeType::Minor]);
    let groups = PackageGroups::new();

    let result = get_disallowed_change_types("foo", &infos, &groups, &None);
    assert_eq!(result, Some(vec![ChangeType::Major, ChangeType::Minor]));
}

// TS: "returns disallowedChangeTypes for package group"
#[test]
fn returns_group_level_disallowed() {
    let infos = make_infos("foo");

    let mut groups = PackageGroups::new();
    groups.insert(
        "grp1".to_string(),
        PackageGroupInfo {
            package_names: vec!["foo".to_string()],
            disallowed_change_types: Some(vec![ChangeType::Major]),
        },
    );

    let result = get_disallowed_change_types("foo", &infos, &groups, &None);
    assert_eq!(result, Some(vec![ChangeType::Major]));
}

// TS: "returns disallowedChangeTypes for package if not in a group"
#[test]
fn returns_package_level_if_not_in_group() {
    let infos = make_infos_with_disallowed("foo", vec![ChangeType::Minor]);

    let mut groups = PackageGroups::new();
    groups.insert(
        "grp1".to_string(),
        PackageGroupInfo {
            package_names: vec!["bar".to_string()],
            disallowed_change_types: Some(vec![ChangeType::Major]),
        },
    );

    let result = get_disallowed_change_types("foo", &infos, &groups, &None);
    assert_eq!(result, Some(vec![ChangeType::Minor]));
}

// TS: "prefers disallowedChangeTypes for group over package"
#[test]
fn prefers_group_over_package() {
    let infos = make_infos_with_disallowed("foo", vec![ChangeType::Minor]);

    let mut groups = PackageGroups::new();
    groups.insert(
        "grp1".to_string(),
        PackageGroupInfo {
            package_names: vec!["foo".to_string()],
            disallowed_change_types: Some(vec![ChangeType::Major]),
        },
    );

    let result = get_disallowed_change_types("foo", &infos, &groups, &None);
    assert_eq!(result, Some(vec![ChangeType::Major]));
}

File: go/internal/changefile/disallowed_change_types_test.go
package changefile_test

import (
	"testing"

	"github.com/microsoft/beachball/internal/changefile"
	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/stretchr/testify/assert"
)

var testRoot = testutil.FakeRoot()

// returns null for unknown package
func TestGetDisallowedChangeTypes_ReturnsNilForUnknownPackage(t *testing.T) {
	infos := types.PackageInfos{}
	groups := types.PackageGroups{}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("unknown-pkg", infos, groups, opts)
	assert.Nil(t, result)
}

// falls back to main option for package without disallowedChangeTypes
func TestGetDisallowedChangeTypes_FallsBackToMainOption(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	groups := types.PackageGroups{}
	opts := &types.BeachballOptions{}
	opts.DisallowedChangeTypes = []string{"major"}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{"major"}, result)
}

// returns disallowedChangeTypes for package
func TestGetDisallowedChangeTypes_ReturnsPackageLevelDisallowedTypes(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	infos["foo"].PackageOptions = &types.PackageOptions{
		DisallowedChangeTypes: []string{"major", "minor"},
	}
	groups := types.PackageGroups{}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{"major", "minor"}, result)
}

// Not possible (Go doesn't distinguish between null and unset):
// returns null if package disallowedChangeTypes is set to null

// returns empty array if package disallowedChangeTypes is set to empty array
func TestGetDisallowedChangeTypes_ReturnsEmptyArrayForEmptyPackageDisallowedTypes(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	infos["foo"].PackageOptions = &types.PackageOptions{
		DisallowedChangeTypes: []string{},
	}
	groups := types.PackageGroups{}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{}, result)
}

// returns disallowedChangeTypes for package group
func TestGetDisallowedChangeTypes_ReturnsGroupLevelDisallowedTypes(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	groups := types.PackageGroups{
		"grp1": &types.PackageGroup{
			Name:                  "grp1",
			Packages:              []string{"foo"},
			DisallowedChangeTypes: []string{"major", "minor"},
		},
	}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{"major", "minor"}, result)
}

// Not possible (Go doesn't distinguish between null and unset):
// returns null if package group disallowedChangeTypes is set to null

// returns empty array if package group disallowedChangeTypes is set to empty array
func TestGetDisallowedChangeTypes_ReturnsEmptyArrayForEmptyGroupDisallowedTypes(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	groups := types.PackageGroups{
		"grp1": &types.PackageGroup{
			Name:                  "grp1",
			Packages:              []string{"foo"},
			DisallowedChangeTypes: []string{},
		},
	}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{}, result)
}

// returns disallowedChangeTypes for package if not in a group
func TestGetDisallowedChangeTypes_ReturnsPackageDisallowedTypesIfNotInGroup(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	infos["foo"].PackageOptions = &types.PackageOptions{
		DisallowedChangeTypes: []string{"patch"},
	}
	groups := types.PackageGroups{
		"grp1": &types.PackageGroup{
			Name:                  "grp1",
			Packages:              []string{"bar"},
			DisallowedChangeTypes: []string{"major", "minor"},
		},
	}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{"patch"}, result)
}

// prefers disallowedChangeTypes for group over package
func TestGetDisallowedChangeTypes_PrefersGroupOverPackage(t *testing.T) {
	infos := testutil.MakePackageInfosSimple(testRoot, "foo")
	infos["foo"].PackageOptions = &types.PackageOptions{
		DisallowedChangeTypes: []string{"patch"},
	}
	groups := types.PackageGroups{
		"grp1": &types.PackageGroup{
			Name:                  "grp1",
			Packages:              []string{"foo"},
			DisallowedChangeTypes: []string{"major", "minor"},
		},
	}
	opts := &types.BeachballOptions{}

	result := changefile.GetDisallowedChangeTypes("foo", infos, groups, opts)
	assert.Equal(t, []string{"major", "minor"}, result)
}
